### DISCUSSION TITLE
How to write unit tests for validator?

### DISCUSSION THREAD

**Discussion by durianwaffle** [eclipse-langium/langium#1684] [Q&A] [answered]
I am writing unit test for validator in langium.
I am wondering how do I pass the `validationAcceptor` into my validationFunction `checkAssignmentType`. Any example on it?
I saw a similar example but still not clear how to text validator since there all my validationFunctions return void type.
![image](https://github.com/user-attachments/assets/e4c9a8bf-a4fa-48d0-8e55-e0e347cfe599)

https://github.com/eclipse-langium/langium/blob/main/examples/requirements/src/language-server/tests-lang-validator.ts

**Comment by msujew** [comment, marked as answer]
Hey @durianwaffle,

we export `expectError, expectIssue, expectNoIssues, expectWarning, validationHelper` from `langium/test` that should help you to write validation tests. See also our own validation tests:

https://github.com/eclipse-langium/langium/blob/main/packages/langium/test/grammar/grammar-validator.test.ts

**Reply by durianwaffle** [reply]
Thank you so much for your help! Now it works

**Reply by durianwaffle** [reply]
Hi @msujew , do you know how to debug for validator?

**Reply by msujew** [reply]
Are you talking about debugging the tests or the language server?

**Reply by durianwaffle** [reply]
I am asking about debugging validator in language server. I think unite test can console log out the result to check😃

**Reply by msujew** [reply]
If you've generated your project with the yeomam generator, there should be a `Attach to language server` debug config available that you can use to debug all language server code (including the validator).

**Reply by durianwaffle** [reply]
Okay thank you @msujew

**Reply by durianwaffle** [reply]
I have another question, if I used other services in validation provider, how to add them into UT?

In the example which used `validationHelper`, there is no place to inject other services.
```{TypeScript}
const validateFeatureCall = validationHelper<MyDSLGrammarAST.AssignmentStatement>(services.MyDSL);
```

**Reply by msujew** [reply]
Doesn't the `services.MyDSL` instance already contain all injected services?

**Reply by durianwaffle** [reply]
Yes you are right. But I still can't import all the services using validationHelper. For `hoverProvider` and `completionProvider`, I can easily inject and use other services because they are all classes. But `validationChecks` is a `function`. How do I use other services in UT for validator?

```{TypeScript}
export function registerValidationChecks(services: MyDSLServices) {
    const registry = services.validation.ValidationRegistry;
    const validator = services.validation.MyDSLValidator;
    const checks: ValidationChecks<MyDSLAstType> = {
        AssignmentStatement: validator.checkAssignmentType,
        FeatureCall: [
            validator.checkFeaturecallDefined,
            validator.checkFeatureCallArgumentsNumberMatchParameterNumber,
        ]
    };
    registry.register(checks, validator);
}
```

**Reply by msujew** [reply]
You can inject services in your `MyDSLValidator` class constructor.

**Reply by msujew** [reply]
I really cannot follow. What do you want to achieve? Can you show some (pseudo)-code for what you want to do?

**Reply by durianwaffle** [reply]
I still face issues using validationHelper to inject my own services. I have injected services in each constructor of each of my services and validator.
So, I used constructor to **new a validator and check the content of ValidationAcceptor** in it.
But I found that if I use `expect(mockValidationAcceptor).toHaveBeenCalledWith('error', 'This is an error message', expect.anything())`, there will be **JS out of memory issue**. Then I used `expect(mockValidationAcceptor).toHaveBeenCalledTimes(1);` which works. But even when it doesn't have validation error the mockValidationAcceptor will be called

**Reply by msujew** [reply]
How do you create the `mockValidationAcceptor`?

**Reply by durianwaffle** [reply]
I created a `mockValidationAcceptor` like the one shown below.
```typescript
mockValidationAcceptor = vi.fn(
        <N extends AstNode>(
            severity: 'error' | 'warning' | 'info' | 'hint',
            message: string,
            info: DiagnosticInfo<N>
        ) => {
            return message;
        }
    )
```
But now I know that I should use the code below instead in beforeAll. It will work.
```typescript
const acceptor = () => null;
toSpyOn = {
    acceptor
};
acceptorSpy = vitest.spyOn(toSpyOn, 'acceptor');
validator = new MyDSLValidator(aService, bService);
validatorAny = validator as any;
validatorAny['checkFeatureCallArguments'](featureCall, toSpyOn.acceptor);
```
And always use `acceptorSpy.mockClear();` in each `test` to make sure the call times is not accumulated.
